bundles/auth/org.eclipse.smarthome.auth.oauth2client/src/main/java/org/eclipse/smarthome/auth/oauth2client/internal/OAuthStoreHandlerImpl.java

```diff
@@ -0,0 +1,531 @@
+
+            // Need to override the date serialization because the milli-seconds are truncated if not handled properly
+            gson = new GsonBuilder()
+                    .registerTypeAdapter(Date.class,
+                            (JsonDeserializer<Date>) (json, typeOfT,
+                                    context) -> new Date(json.getAsJsonPrimitive().getAsLong()))
+                    .registerTypeAdapter(Date.class, (JsonSerializer<Date>) (date, type,
+                            jsonSerializationContext) -> new JsonPrimitive(date.getTime()))
+                    .create();
+        }
+
+        @SuppressWarnings("unchecked")
+        public HashSet<String> getAllHandlesFromIndex() {
+            HashSet<String> handlesFromStoreageIndex = new HashSet<>();
+            try {
+                String allHandlesStr = get(STORE_KEY_INDEX_OF_HANDLES);
+                logger.debug("All available handles: {}", allHandlesStr);
+                if (allHandlesStr == null) {
+                    return handlesFromStoreageIndex;
+                }
+                return gson.fromJson(allHandlesStr, HashSet.class);
+            } catch (RuntimeException storeNotAvailable) {
+                return handlesFromStoreageIndex; // empty
+            }
+        }
+
+        public String get(@NonNull String key) {
+            try {
+                storageLock.lock();
+                if (_storage == null) {
+                    throw new RuntimeException(EXCEPTION_STORAGE);
+                }
+                return _storage.get(key);
+            } finally {
+                storageLock.unlock();
+            }
+        }
+
+        public Object get(String handle, StorageRecordType recordType) {
+            try {
+                storageLock.lock();
+                if (_storage == null) {
+                    throw new RuntimeException(EXCEPTION_STORAGE);
+                }
+
+                String value = _storage.get(recordType.getKey(handle));
+                if (value == null) {
+                    return null;
+                }
+
+                // update last used when it is an access token
+                if (recordType.equals(ACCESS_TOKEN_RESPONSE)) {
+                    AccessTokenResponse accessTokenResponse = gson.fromJson(value, AccessTokenResponse.class);
+                    return accessTokenResponse;
+                } else if (recordType.equals(SERVICE_CONFIGURATION)) {
+                    PersistedParams params = gson.fromJson(value, PersistedParams.class);
+                    return params;
+                } else if (recordType.equals(LAST_USED)) {
+                    Date lastUsedDate = gson.fromJson(value, Date.class);
+                    return lastUsedDate;
+                }
+                return null;
+            } finally {
+                storageLock.unlock();
+            }
+        }
+
+        public void put(String handle, Long lastUsed) {
+            try {
+                storageLock.lock();
```

Comment:
It is a recommended practice to lock the monitor outside try block. If you use it inside and if the lock fails, the finally will be invoked and it will again try to unlock the monitor which has never been locked.